feat: update C/igraph

Antonov548 opened the pull request.

Antonov548 commented:
@szhorvat What is the diference between new type `ALL_VERTEX_QTY` and `VERTEX_QTY`?

szhorvat commented:
It was added by @krlmlr, it's best if he comments.

But as I remember:

 - If there is a `VERTEX_SELECTOR vids`, we use `VERTEX_QTY`, and make it depend on `vids`. This helps Stimulus generate appropriate code, such as adding vertex names to the result vector.
 - Kirill added `ALL_VERTEX_QTY` to make the same possible when the function returns a result for each vertex, without the possibility to subset vertices using a `VERTEX_SELECTOR`. This is new and I _think_ the Stimulus-side support is not ready yet. As I recall, previously these were just a `VECTOR`, but that doesn't encode the information that the values belong to vertices.

szhorvat commented:
Can you please pick up the latest commit? Two new functions were added in C since this PR was started.

szhorvat commented on R/aaa-auto.R:
```diff
@@ -330,6 +330,19 @@ simple_interconnected_islands_game_impl <- function(islands.n, islands.size, isl
   res
 }
 
+chung_lu_game_impl <- function(expected.out.deg, expected.in.deg=NULL, loops=TRUE, variant=ORIGINAL) {
```
There needs to be translation code for `igraph_chung_lu_t` values.

Antonov548 replied:
Does it make sense to make it in seperate PR?

szhorvat replied:
I think they're might be warnings if `=ORIGINAL` is left in there. I suggest you disable auto generation for this function at the moment, then we can handle it in a later PR

Antonov548 commented:
> mmit? Two new functions

Updated to latest. For now I have added `ALL_VERTEX_QTY` same as `VERTEX_QTY`.
2 tests are failing. @szhorvat is it something that expected to be broken? 

```bash
Failure (test-all.st.cuts.R:31:3): all.st.cuts works
`x` (`actual`) not equal to `expected` (`expected`).

`actual` is length 3
`expected` is length 4

  `actual[[2]]`: 1 3          
`expected[[2]]`: 1 2 9 8 7 6 5

  `actual[[3]]`: 1 3 2 9 8 7 6 5
`expected[[3]]`: 1 3            

`actual[[4]]` is absent
`expected[[4]]` is a double vector (1, 3, 2, 9, 8, ...)
```

szhorvat commented:
I'm outside, commenting on my phone. There was a major bugfix for this function but I thought that was already released with R.

Could you give me the last C commit that was used in R? Then I'll take a look

szhorvat commented:
Please be sure to update to the latest C/igraph `master` as of this writing. I just fixed a serious bug I introduced a few days ago—let's not pick that up for R! See https://github.com/igraph/igraph/commit/2f5ffca89c7566ed539d866ccb2c9f4a1e822a8d for the fix.

szhorvat commented:
@Antonov548 Can you please verify manually if the new/old results from that test are actually correct? See the documentation for what `cuts` and `partition1s` are.

Antonov548 commented:
> @Antonov548 Can you please verify manually if the new/old results from that test are actually correct? See the documentation for what `cuts` and `partition1s` are.

Updated with your latest bug fixes. 
I will try to check it manually.

szhorvat commented:
Here's a nice plot of the graph to make verification easier:

<img width="360" alt="image" src="https://github.com/igraph/rigraph/assets/1212871/b67411aa-46a9-4071-a462-80ae8b411c50">

szhorvat commented:
OK, so in the current version (before this PR), I'm getting:

```
> cc$cuts
[[1]]
+ 2/14 edges from 97a0acc (vertex names):
[1] s->a s->b

[[2]]
+ 2/14 edges from 97a0acc (vertex names):
[1] s->b a->t

[[3]]
+ 2/14 edges from 97a0acc (vertex names):
[1] s->a b->t

[[4]]
+ 2/14 edges from 97a0acc (vertex names):
[1] a->t b->t

> cc$partition1s
[[1]]
+ 1/9 vertex, named, from 97a0acc:
[1] s

[[2]]
+ 7/9 vertices, named, from 97a0acc:
[1] s a 5 4 3 2 1

[[3]]
+ 2/9 vertices, named, from 97a0acc:
[1] s b

[[4]]
+ 8/9 vertices, named, from 97a0acc:
[1] s b a 5 4 3 2 1

> 
```

Clearly, `s->b, a->t` is not a valid cut as the path `s->a->3->b->t` remains between `s` and `t`.

After this PR the result is correct and contains only 3 cuts, namely `s->a s->b`, `s->a b->t`, `a->t b->t`. This line in the test expects four (incorrectly):

```R
expect_that(unvs(cc$cuts), equals(list(c(1, 2), c(2, 3), c(1, 9), c(3, 9))))
```

Why does this line not fail? Do you understand this @maelle ? Does R report only the last test failure?

szhorvat commented:
@Antonov548 Either way, the conclusion is that:

 - The test was invalid, and the expected result should be updated
 - The current released version is still based on a buggy version of the C core. Many fixes have been made since then. Please include a C core update in the new release @krlmlr 
 - @Antonov548, this PR looks fine to me! The only needed change is to update this failing test

Antonov548 commented:
> @Antonov548 Either way, the conclusion is that:
> 
> * The test was invalid, and the expected result should be updated
> * The current released version is still based on a buggy version of the C core. Many fixes have been made since then (the last two C releases were very bugfix-heavy). Please include a C core update in the new release @krlmlr. Do expect revdepchecks to fail due to the bugfix—but this is a good thing!
> * @Antonov548, this PR looks fine to me! The only needed change is to update this failing test

Thanks for better visualization and also for checking it.
I have checked on the visualization that I have attached and seems currently it works fine.

So, I will just update the tests.
Also, for me this tests are failing.

![image](https://github.com/igraph/rigraph/assets/22891541/c522ddbc-14e4-4c7f-b787-0548e1f4d7f3)

szhorvat commented:
Can you please update to the latest commit again? I made some changes which I'd like to be tested using the R/igraph test suite.

szhorvat commented:
Sorry to keep asking for updates, but I just made another bugfix. Perhaps just update once more before this is actually merged?

Antonov548 commented:
> Sorry to keep asking for updates, but I just made another bugfix. Perhaps just update once more before this is actually merged?

Updated. Not sure what happened with Coverage check. The error in the logs is not informative.

maelle commented:
@Antonov548 locally I get this error when running the tests:

```r
Error (test-authority.score.R:94:5): authority_score survives stress test
Error in `hub_score_impl(graph = graph, scale = scale, weights = weights, 
    options = options)`: At vendor/cigraph/src/linalg/arpack.c:1102 : ARPACK error, No shifts could be applied during a cycle of the Implicitly restarted Arnoldi iteration. One possibility is to increase the size of NCV relative to NEV
Backtrace:
    ▆
 1. └─igraph::hub_score(G) at test-authority.score.R:94:5
 2.   └─igraph:::hub_score_impl(...) at rigraph/R/centrality.R:1146:3
```

maelle commented:
maybe it'd be better after #1352

Antonov548 commented:
> @Antonov548 locally I get this error when running the tests:
> 
> ```r
> Error (test-authority.score.R:94:5): authority_score survives stress test
> Error in `hub_score_impl(graph = graph, scale = scale, weights = weights, 
>     options = options)`: At vendor/cigraph/src/linalg/arpack.c:1102 : ARPACK error, No shifts could be applied during a cycle of the Implicitly restarted Arnoldi iteration. One possibility is to increase the size of NCV relative to NEV
> Backtrace:
>     ▆
>  1. └─igraph::hub_score(G) at test-authority.score.R:94:5
>  2.   └─igraph:::hub_score_impl(...) at rigraph/R/centrality.R:1146:3
> ```

Is it reason why Coverage is failing? 
Maybe then it's really better to check again after removing deprecated functions.

maelle commented:
no that's a different error apparently. so yeah let's wait until the other PR is merged and then rebase :grin:

szhorvat commented:
I might want to submit a feature PR (new function) that requires the C core update. Should I make a PR against this PR, or wait, @krlmlr ? It's not a huge deal if we don't make it, this should NOT block the release.

szhorvat commented:
> @Antonov548 locally I get this error when running the tests:
> 
> ```r
> Error (test-authority.score.R:94:5): authority_score survives stress test
> Error in `hub_score_impl(graph = graph, scale = scale, weights = weights, 
>     options = options)`: At vendor/cigraph/src/linalg/arpack.c:1102 : ARPACK error, No shifts could be applied during a cycle of the Implicitly restarted Arnoldi iteration. One possibility is to increase the size of NCV relative to NEV
> Backtrace:
>     ▆
>  1. └─igraph::hub_score(G) at test-authority.score.R:94:5
>  2.   └─igraph:::hub_score_impl(...) at rigraph/R/centrality.R:1146:3
> ```

This is why I asked to run the R test suite with the C core update. Dealing with this is extremely frustrating. You fix one case, it breaks the other. You fix it on one platform, it's broken on the other. ARPACK is fickle and will never work for all cases.

 - You get this locally only, correct? Not in the CI.
 - @Antonov548, do you also get it locally?
 - @maelle What platforms are you on? What OS, what R version, what CPU architecture?

szhorvat commented:
> maybe it'd be better after #1352

No, this won't fix it. It's due to recent changes I made in the C core. We might need to end up modifying tests ... But first I'd like to know your answers to my questions from the previous comment.

Antonov548 commented:
> > @Antonov548 locally I get this error when running the tests:
> > ```r
> > Error (test-authority.score.R:94:5): authority_score survives stress test
> > Error in `hub_score_impl(graph = graph, scale = scale, weights = weights, 
> >     options = options)`: At vendor/cigraph/src/linalg/arpack.c:1102 : ARPACK error, No shifts could be applied during a cycle of the Implicitly restarted Arnoldi iteration. One possibility is to increase the size of NCV relative to NEV
> > Backtrace:
> >     ▆
> >  1. └─igraph::hub_score(G) at test-authority.score.R:94:5
> >  2.   └─igraph:::hub_score_impl(...) at rigraph/R/centrality.R:1146:3
> > ```
> 
> This is why I asked to run the R test suite with the C core update. Dealing with this is extremely frustrating. You fix one case, it breaks the other. You fix it on one platform, it's broken on the other. ARPACK is fickle and will never work for all cases.
> 
> * You get this locally only, correct? Not in the CI.
> * @Antonov548, do you also get it locally?
> * @maelle What platforms are you on? What OS, what R version, what CPU architecture?
> * If you are using Linux, or a non-CRAN R distribution, what BLAS/LAPACK do you have (version, implementation)?

No, I don't have it locally. Ubuntu 22.04.2 LTS, R 4.3.0

szhorvat commented:
@Antonov548 @maelle Can you update to the latest commit again and test if that ARPACK error is still there?

szhorvat commented:
@Antonov548 Sigh ... now something else got broken. Can you update once more please, and also let me know the current update procedure so I can do this next time on my own?

szhorvat commented:
It'd be good to fix the scripts in `tools` at some point. `tools/update-cigraph.sh` does not work at the moment.

szhorvat commented:
I think the hub/authority scores related changes are fine as they are now, even if failures remain on a some platforms. We can look at those failures later, but with enough random fuzzing, some failure is basically guaranteed to come out. This is not new (see e.g. https://github.com/igraph/igraph/issues/1990 and https://github.com/igraph/igraph/issues/1641). I expect the random seed used in this test was just lucky.  Note that the affected test is skipped on CRAN, for good reason.

maelle commented:
> @maelle What platforms are you on? What OS, what R version, what CPU architecture?
> If you are using Linux, or a non-CRAN R distribution, what BLAS/LAPACK do you have (version, implementation)?

- Ubuntu 20.04.6 LTS
- R 4.2.0 (I actually plan to update soon, what a shame I'm still using that)

How do I better print the BLAS/LAPACK version?

maelle commented:
Currently re-running the tests with a more recent R version :innocent:

maelle commented:
No error!! Only a warning

```r
Warning (test-make_graph.R:58:3): make_empty_graph gives an error for invalid arguments
NAs introduced by coercion
Backtrace:
    ▆
 1. ├─testthat::expect_error(make_empty_graph("spam")) at test-make_graph.R:58:3
 2. │ └─testthat:::expect_condition_matching(...)
 3. │   └─testthat:::quasi_capture(...)
 4. │     ├─testthat (local) .capture(...)
 5. │     │ └─base::withCallingHandlers(...)
 6. │     └─rlang::eval_bare(quo_get_expr(.quo), quo_get_env(.quo))
 7. └─igraph::make_empty_graph("spam")
 ```

szhorvat commented:
> Currently re-running the tests with a more recent R version 😇

Let's move ahead then and merge this ASAP.

Antonov548 commented:
@szhorvat after merge with main branch tests for `authority_score()` doesn't work anymore.
Please take a look at the difference of results. It looks wierd and it's only after merge. In this branch and main seperatly it works fine.

But also this function is deprecated and use `hit_scores()` in pull request with removing deprecated function. Should we just skip tests in this PR and enable again in pull request where deprecated function are removed?

```bash
Failure (test-authority.score.R:63:3): authority scores of a ring are all one
authority_score(g3)$vector (`actual`) not equal to rep(1, vcount(g3)) (`expected`).

     actual     | expected                  
 [1] 0.99999231 - 1.00000000 [1]            
 [2] 1.00000000 | 1.00000000 [2]            
 [3] 0.99999231 - 1.00000000 [3]            
 [4] 1.00000000 | 1.00000000 [4]            
 [5] 0.99999231 - 1.00000000 [5]            
 [6] 1.00000000 | 1.00000000 [6]            
 [7] 0.99999231 - 1.00000000 [7]            
 [8] 1.00000000 | 1.00000000 [8]            
 [9] 0.99999231 - 1.00000000 [9]            
[10] 1.00000000 | 1.00000000 [10] 
```

szhorvat commented:
The merge overwrote the fix I made in https://github.com/igraph/rigraph/pull/1378/commits/e7705d845590db29d6245702d800756474d125cc  Please re-do the fix to this test.

Antonov548 commented:
> The merge overwrote the fix I made in [e7705d8](https://github.com/igraph/rigraph/commit/e7705d845590db29d6245702d800756474d125cc) Please re-do the fix to this test.

Sorry, didn't notice this change during merge. Fixed.

szhorvat commented:
Please don't skip the comments I added, otherwise someone will likely break this again.

Antonov548 commented:
> Please don't skip the comments I added, otherwise someone will likely break this again.

Yes, I added comemnt in seperate commit.

maelle commented:
I fetched the PR.

Running the tests yields

```r
Warning (test-make_graph.R:58:3): make_empty_graph gives an error for invalid arguments
NAs introduced by coercion
Backtrace:
    ▆
 1. ├─testthat::expect_error(make_empty_graph("spam")) at test-make_graph.R:58:3
 2. │ └─testthat:::expect_condition_matching(...)
 3. │   └─testthat:::quasi_capture(...)
 4. │     ├─testthat (local) .capture(...)
 5. │     │ └─base::withCallingHandlers(...)
 6. │     └─rlang::eval_bare(quo_get_expr(.quo), quo_get_env(.quo))
 7. └─igraph::make_empty_graph("spam")
────────────────────────────────────
```
I opened an issue about this: #1400

I'll run covr locally.

szhorvat commented:
Is there any reason to keep this unmerged? It's an inconvenience and makes it impossible to submit PRs that add new functionality. Coverage is broken everywhere, including `main`, anyway.

Let's please not let bureaucratic reasons (such as always wanting to use Aviator, which requires checks to pass) stand in the way of efficiency.

I'd like to suggest merging ASAP (ideally today), and running revdepchecks. In this particular case, resolving true revdepcheck issues is for the C core, not R, so it shouldn't be a big concern. I don't expect the C core upgrade to have introduced new bugs, but it may cause revdepcheck failures for packages that relied on buggy behaviour (see the test that's updated in this PR).

CC @krlmlr

krlmlr commented:
Running revdepchecks on this branch now.

szhorvat commented:
Great, please ping me when you have the results!

szhorvat commented:
Neither of these two failures are related to the update. 

The `netropy` one is a problem with that package, and should have occurred earlier too, already in 2.0.0. Maybe it didn't because they have some non-deterministic behaviour in their test?

The `tidygraph` one must be related to changes on `main` that were merged?

krlmlr commented:
Both failures are present in the main branch too. Let's merge when green, if Aviator fails to do this.

krlmlr reviewed (approved):
Thanks for pushing this effort!
